charging_station.py

```diff
@@ -0,0 +1,155 @@
+        self.nb_fast = 2
+        self.pmax_fast = 22
+        self.pmax_slow = 3
+        self.cmax = 40*4 # Maximal capacity of the CS when the 4 slots are used
+        self.depart = {"slow" : np.zeros(2), "fast" : np.zeros(2)} # Time of departure of every cars
+        self.arrival = {"slow" : np.zeros(2), "fast" : np.zeros(2)} # Time of arrival of every cars
+        self.here = {"slow" : np.ones(2), "fast" : np.ones(2)}
+        self.imbalance=[]
+
+    def update_battery_stock(self,time,load_battery):
+        self.nb_cars(time) # We check what cars is here
+        p_max = {"slow" : [3*self.here["slow"][0],3*self.here["slow"][1]], "fast" : [22*self.here["fast"][0],22*self.here["fast"][1]]}
+        c_max = {"slow" : [40*self.here["slow"][0],40*self.here["slow"][1]], "fast" : [40*self.here["fast"][0],40*self.here["fast"][1]]}
+        # p_max and c_max depend on whether the car is here or not.
+        for speed in ["slow","fast"] :
+            for i in range(2):
+                if abs(load_battery[speed][i]) >= p_max[speed][i]:
+                    load_battery[speed][i] = p_max[speed][i]*np.sign(load_battery[speed][i])
+            # Can't put more power than p_max
+            new_stock = { "slow" : [0,0], "fast" : [0,0] }
+
+            new_stock["slow"][0] = self.battery_stock["slow"][time][0] + (self.efficiency*max(0,load_battery["slow"][0])+min(0,load_battery["slow"][0])/self.efficiency)*self.dt
+            new_stock["slow"][1] = self.battery_stock["slow"][time][1] + (self.efficiency*max(0,load_battery["slow"][1])+min(0,load_battery["slow"][1])/self.efficiency)*self.dt
+            # We update the new stock of each batteries "slow"
+
+            new_stock["fast"][0]=self.battery_stock["fast"][time][0] + (self.efficiency*max(0,load_battery["fast"][0])+min(0,load_battery["fast"][0])/self.efficiency)*self.dt
+            new_stock["fast"][1]=self.battery_stock["fast"][time][1] + (self.efficiency*max(0,load_battery["fast"][1])+min(0,load_battery["fast"][1])/self.efficiency)*self.dt
+            # We update the new stock of each batteries "fast"
+
+
+        for speed in ["slow","fast"] :
+            for i in range(2):
+                if new_stock[speed][i] < 0:
+                    load_battery[speed][i] = -(self.battery_stock[speed][time][i])/(self.efficiency*self.dt)
+                    new_stock[speed][i] = 0
+            # We can't discharge the batterie under 0
+
+                elif new_stock[speed][i] > c_max[speed][i]:
+                    load_battery[speed][i] = (c_max[speed][i] - self.battery_stock[speed][time][i] ) / (self.efficiency*self.dt)
+                    new_stock[speed][i] = c_max[speed][i]
+            # We can't charge the batteries over their maximum capacities
+
+        for speed in ["slow","fast"] :
+            for i in range(2):
+                if abs(load_battery[speed][i]) >= p_max[speed][i] :
+                    load_battery[speed][i] = p_max[speed][i]*np.sign(load_battery[speed][i])
```

Comment:
I think that this verification is not useful here given that it has been done above (L36) and that the absolute value of the load can only be reduced from this Line (in case when the bounds of the battery are violated)

Reply:
A priori si le cas des VEs non présents est géré de manière séparée je pense que cette seconde vérification sera bien obsolète